File: Algorithm/Go/998. Maximum Binary Tree II.go
/*
 * Created on Tue Feb 26 2019 21:27:6
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// recursive, always insert to the right subtree of the current root
func insertIntoMaxTree(root *TreeNode, val int) *TreeNode {
	if root == nil || root.Val < val {
		new_root := new(TreeNode)
		new_root.Val = val
		new_root.Left = root
		return new_root
	} else {
		root.Right = insertIntoMaxTree(root.Right, val)
		return root
	}
}

File: Algorithm/Python/998. Maximum Binary Tree II.py
# -*- coding: utf-8 -*-
# Created on Tue Feb 26 2019 21:5:55
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# recursive, always insert to the right subtree of the current root
class Solution(object):
    def insertIntoMaxTree(self, root, val):
        """
        :type root: TreeNode
        :type val: int
        :rtype: TreeNode
        """
        if root is None or val > root.val:
            new_root = TreeNode(val)
            new_root.left = root
            return new_root
        else:
            root.right = self.insertIntoMaxTree(root.right, val)
            return root